server/src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -725,7 +725,6 @@ public final <T extends TransportResponse> void sendRequest(
             return;
         } catch (Exception exception) {
             // shouldn't happen in practice, but handle it anyway to be sure
-            assert false : exception;
```

Comment:
Hi @DaveCTurner , I removed this in order to be able to do the `TransportBroadcastUnpromotableActionTests#testInvalidNodes` test. I think we said to test if there were invalid nodes in the index shard routing table. But the test then trips here. If this branch cannot happen in practice, then I guess I will remove the test.

Reply:
ah I think I see. Unknown nodes should yield something like a `NodeNotConnectedException` which would be handled by the other catch block.

Reply:
@DaveCTurner , true about the other catch block, but in my case the testInvalidNodes results in a `java.lang.AssertionError: java.lang.NullPointerException: discovery node must not be null`. Because `TransportBroadcastUnpromotableAction#doExecute` does `final DiscoveryNode node = clusterState.nodes().get(shardRouting.currentNodeId());` (with a wrong node id due to the test) and this results in null which is then passed to the sendRequest method whose `isLocalNode()` requires a non-null node id.

Reply:
Yes, I think we should map this to a `NodeNotConnectedException` somehow. This is definitely possible because `clusterState.nodes()` comes from a different cluster state from the ISRT so may indeed be missing nodes.

Reply:
OK I see, I did that, thanks!